src/dune_pkg/archive_type.ml

```diff
@@ -2,18 +2,45 @@ open Stdune
 module Process = Dune_engine.Process
 open Fiber.O
 
-let bin =
-  lazy
-    (match Bin.which ~path:(Env_path.path Env.initial) "tar" with
-     | Some x -> x
-     | None -> Dune_engine.Utils.program_not_found "tar" ~loc:None)
+type t =
+  { bin : Path.t Lazy.t
+  ; suffixes : string list
+    (* Identify a file by its suffix rather than by its extension, since
+       ".tar.gz" files only have an extension of ".gz". *)
+  ; build_args : archive:Path.t -> target_in_temp:Path.t -> string list
+  }
+
+let make ~bin_name ~suffixes ~build_args =
+  let bin =
+    lazy
+      (match Bin.which ~path:(Env_path.path Env.initial) bin_name with
+       | Some x -> x
+       | None -> Dune_engine.Utils.program_not_found bin_name ~loc:None)
+  in
+  { bin; suffixes; build_args }
+;;
+
+let tar =
+  make
+    ~bin_name:"tar"
+    ~suffixes:[ ".tar"; ".tar.gz"; ".tgz"; ".tar.bz2"; ".tbz" ]
```

Comment:
Come to think of it, support for `.gz` is dependant on `gunzip` (which is admittedly so common we can probably assume that everyone with `tar` has it) but `bunzip2` is getting rarer these days and according to [this issue](https://github.com/ocaml/opam/issues/6171) OPAM also supports `unxz`.

(As far as I know GNU `tar` is just shelling out to those internally, libarchive's `bsdtar` might actually decompress them using libarchive)

Reply:
If we separate the questions of "What command does dune need to run to extract an archive?" and "What programs need to be in PATH to extract an archive?", dune currently only handles the first, and will fail with an (often helpful) error if one of the required programs is missing. We could try to improve the ergonomics of this by making it more clear what external tools are needed to handle each type of archive, but that should happen in a different change from this one.

Reply:
Ok, once this is merged I'll probably look into this. That way we can also untangle `tar` from `gzip`/`bzip` or whatever other compressor is used.